src/Device/Driver/FANET/Parser.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2016 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "Device.hpp"
+#include "NMEA/Info.hpp"
+#include "Parser.hpp"
+#include "FANET/FanetAddress.hpp"
+
+/*
+ * Updates existing station or allocates new one
+ * Returns false if the list is full, true otherwise
+ */
+bool
+FanetDevice::AddStationToList(StationsList &stations_list, double clock,
+                              const FanetStation &station) {
+  FanetStation *fanet_slot = stations_list.FindStations(station.address);
+
+  if (fanet_slot == nullptr) {
+    fanet_slot = stations_list.AllocateStation();
+    if (fanet_slot == nullptr)
+      // no more slots available
+      return false;
+
+    fanet_slot->Clear();
+    fanet_slot->address = station.address;
+    fanet_slot->location = station.location;
+    fanet_slot->location_available = station.location_available;
+  }
+
+  fanet_slot->valid.Update(clock);
+  fanet_slot->Update(station);
+
+  return true;
+}
+
+bool
+FanetDevice::IntegrityCheck(char *cmd_str) {
+  for(char *ptr = cmd_str; *ptr != '\0'; ptr++)
+  {
+    if((*ptr >= '0' && *ptr <= '9') ||
+       (*ptr >= 'A' && *ptr <= 'F') ||
+       (*ptr >= 'a' && *ptr <= 'f') ||
+       (*ptr == ',')) {
+      continue;
+    }
+    return false;
+  }
+  return true;
+}
+
+char *
+FanetDevice::RemoveExtraChars(char *cmd_str) {
+  // Remove \r\n and any spaces
+  char *ptr = strchr(cmd_str, '\r');
+  if(ptr == nullptr)
+    ptr = strchr(cmd_str, '\n');
+  if(ptr != nullptr)
```

Comment:
This will preserve all `\n` if at least one `\r` exists. Is that what you want?

Reply:
This will insert \0 in place of first occurance of the \r, if \r is not found looks for \n and inserts it there. That sounds reasonable to me.

Reply:
I'm not convinced, this sounds pretty broken. *If* you believe it is necessary to stop parsing at the next `\r` or `\n`, then leaving `\n` in there under certain conditions just isn't right.
But note that you were already given a single line to parse. So what's the point of this function?